packages/pyright-internal/src/analyzer/typeEvaluator.ts

```diff
@@ -2135,7 +2135,12 @@ export function createTypeEvaluator(importLookup: ImportLookup, evaluatorOptions
                         // If there was no `__init__` or the only `__init__` that was found
                         // was form the `object` class, see if we can find a better `__new__`
                         // method.
-                        if (!methodType || isObjectInit || isSkipConstructor) {
+                        if (
+                            !methodType ||
+                            isObjectInit ||
+                            isSkipConstructor ||
+                            (methodType && isFunction(methodType) && (FunctionType.hasDefaultParameters(methodType) || methodType.details.parameters.length === 0))
```

Comment:
it seems this function `function getCallSignatureInfo()` is only called from completionProvider and signatureHelpProvider.  I was wondering if we should take it out of the typeEvaluator code?

Reply:
Ah, I didn't notice that this was in `getCallSignatureInfo`. Then it shouldn't affect any type evaluations, and I'm less concerned about the change.

When I originally wrote `getCallSignatureInfo`, it needed to be part of the typeEvaluator because it relied on functions that were not exported from the typeEvaluator. We've since exported a lot more functions, so it might be possible to implement it outside of the typeEvaluator. Not sure if it's worth moving though.

Reply:
the logic is saying if __init__ is the following:

`def __init__(*args, **kwargs)`
`def __init__()`

then check for __new_  for a better constructor signature.  this was the case for django stubs, where they had a detailed `__new__` in a child class, but defaultParams for `__init__` in a parent base class.

there is also a way to skip the `isObjectInit` check. currently we call 

`methodType = getBoundMethod(subtype, '__init__');` above
and inside that it doesn't exclude methods from Object, when it makes a call to `lookUpClassMember`